diff --git a/Client.py b/Client.py
index 6e86590..335fd68 100644
--- a/Client.py
+++ b/Client.py
@@ -17,7 +17,6 @@ from .DataHandler import (
     restore_originals,
     restore_song_list,
     find_linked_numbers,
-    get_dict
 )
 
 from CommonClient import (
@@ -131,7 +130,7 @@ class MegaMixContext(CommonContext):
             self.enable_all_diff = self.options["enableAllDiff"]
             self.leeks_needed = self.options["leekWinCount"]
             self.grade_needed = int(self.options["scoreGradeNeeded"]) + 2  # Add 2 to match the games internals
-            self.modData = get_dict(self.options["modData"], True)
+            self.modData = json.loads(self.options["modData"]) if self.options["modData"] else None
             if self.modData:
                 self.modded = True
                 self.mod_pv_list = generate_modded_paths(self.modData, self.path)
diff --git a/DataHandler.py b/DataHandler.py
index 6368ac6..283ef05 100644
--- a/DataHandler.py
+++ b/DataHandler.py
@@ -1,4 +1,5 @@
 import json
+import yaml
 import pkgutil
 import re
 import os
@@ -7,7 +8,6 @@ import sys
 import Utils
 from .SymbolFixer import fix_song_name
 from collections import defaultdict
-import ast
 from typing import Any
 
 
@@ -384,7 +384,7 @@ def extract_mod_data_to_json() -> list[Any]:
     search_text = "Hatsune Miku Project Diva Mega Mix+"
 
     # Regex pattern to capture the outermost curly braces content
-    mod_data_pattern = r'megamix_mod_data:\s*(?:#.*\n)?\s*(\{.*\})'
+    mod_data_pattern = r"megamix_mod_data:\s*(?:#.*\n)?\s*('.*')"
 
     # Initialize an empty list to collect all inputs
     all_mod_data = []
@@ -403,13 +403,13 @@ def extract_mod_data_to_json() -> list[Any]:
 
                     # Process each mod_data block
                     for mod_data_content in matches:
+                        mod_data_match = yaml.safe_load(file_content)
+                        mod_data_content = mod_data_match.get("Hatsune Miku Project Diva Mega Mix+", {}).get("megamix_mod_data", '""')
 
-                        data_dict = get_dict(mod_data_content, False)
-
-                        if data_dict == "":
+                        if isinstance(mod_data_content, dict) or not mod_data_content:
                             continue
 
-                        all_mod_data.append(data_dict)
+                        all_mod_data.append(json.loads(mod_data_content))
 
     total = sum(len(pack) for packList in all_mod_data for pack in packList.values())
     print(f"Found {total} songs")
@@ -417,38 +417,13 @@ def extract_mod_data_to_json() -> list[Any]:
     return all_mod_data
 
 
-def get_dict(mod_data, client):
-
-    if client:
-        trimmed_data = str(mod_data[8:-1])  # Slicing to remove first and last character
-        if trimmed_data == "":
-            return None
-
-    else:
-        # Remove the first 2 and last 2 characters
-        trimmed_data = str(mod_data[2:-2])  # Slicing to remove first and last character
-
-        if trimmed_data == "":
-            return ""
-
-    try:
-        data_dict = ast.literal_eval(trimmed_data)
-    except (ValueError, SyntaxError) as e:
-        print(f"Error parsing data: {e}")
-        data_dict = {}
-
-    return data_dict
-
-
 def get_player_specific_ids(mod_data):
-
     song_ids = []  # Initialize an empty list to store song IDs
 
-    if mod_data == "()":
+    if mod_data == "":
         return song_ids
 
-    trimmed_data = str(mod_data[1:-1])  # Slicing to remove first and last character
-    data_dict = ast.literal_eval(trimmed_data)
+    data_dict = json.loads(mod_data)
 
     for pack_name, songs in data_dict.items():
         for song in songs:
diff --git a/MegaMixCollection.py b/MegaMixCollection.py
index 251c6d9..bc36c09 100644
--- a/MegaMixCollection.py
+++ b/MegaMixCollection.py
@@ -90,7 +90,7 @@ class MegaMixCollections:
                                 formatted_difficulties[difficulty_mapping_modded[key]] = difficulties[key]
 
                         for diff, rating in formatted_difficulties.items():
-                            song_name = str.replace(song[0], "/", "'")
+                            song_name = song[0]
                             song_name = fix_song_name(song_name)
                             song_name = song_name + " " + diff
                             item_id = (song_id * 10) + difficulty_mapping.get(diff, "Difficulty not found")
diff --git a/Options.py b/Options.py
index 912ea59..fc08566 100644
--- a/Options.py
+++ b/Options.py
@@ -1,5 +1,5 @@
 from typing import Dict
-from Options import Toggle, Option, Range, Choice, DeathLink, ItemSet, OptionSet, PerGameCommonOptions
+from Options import Toggle, Option, Range, Choice, DeathLink, ItemSet, OptionSet, PerGameCommonOptions, FreeText
 from dataclasses import dataclass
 
 
@@ -229,10 +229,10 @@ class ExcludeSinger(OptionSet):
     default = {}
 
 
-class ModData(OptionSet):
+class ModData(FreeText):
     """If you are using modded songs, paste the string from the DivaJSON Tool here"""
     display_name = "MegaMixModData"
-    default = {}
+    default = ''
 
 
 @dataclass
diff --git a/SymbolFixer.py b/SymbolFixer.py
index 8f835ba..f05c0bd 100644
--- a/SymbolFixer.py
+++ b/SymbolFixer.py
@@ -78,7 +78,7 @@ def replace_non_ascii_with_space(text):
 
 def special_char_removal(text):
     # Remove apostrophes, commas, and quotation marks
-    cleaned_text = text.replace("'", "").replace(",", "").replace('"', "")
+    cleaned_text = text #.replace("'", "").replace(",", "").replace('"', "")
 
     # Replace multiple spaces with a single space
     cleaned_text = re.sub(r'\s+', ' ', cleaned_text)
diff --git a/__init__.py b/__init__.py
index 9b2e091..0783323 100644
--- a/__init__.py
+++ b/__init__.py
@@ -84,7 +84,7 @@ class MegaMixWorld(World):
 
         while True:
             # In most cases this should only need to run once
-            available_song_keys, song_ids = self.mm_collection.get_songs_with_settings(self.options.allow_megamix_dlc_songs, get_player_specific_ids(str(self.options.megamix_mod_data)[7:]), allowed_difficulties, disallowed_singers, lower_diff_threshold, higher_diff_threshold, self.options.always_pick_hardest)
+            available_song_keys, song_ids = self.mm_collection.get_songs_with_settings(self.options.allow_megamix_dlc_songs, get_player_specific_ids(self.options.megamix_mod_data.value), allowed_difficulties, disallowed_singers, lower_diff_threshold, higher_diff_threshold, self.options.always_pick_hardest)
 
             # Choose victory song from current available keys, so we can access the song id
             chosen_song_index = random.randrange(0, len(available_song_keys))
@@ -325,6 +325,6 @@ class MegaMixWorld(World):
             "leekWinCount": self.get_leek_win_count(),
             "scoreGradeNeeded": self.options.grade_needed.value,
             "autoRemove": bool(self.options.auto_remove_songs),
-            "modData": str(self.options.megamix_mod_data),
+            "modData": self.options.megamix_mod_data.value,
             "enableAllDiff": bool(self.options.enable_all_diff),
         }
