diff --git a/Client.py b/Client.py
index 1c00ecb..7b1cce2 100644
--- a/Client.py
+++ b/Client.py
@@ -63,8 +63,9 @@ class MegaMixContext(CommonContext):
 
         self.game = "Hatsune Miku Project Diva Mega Mix+"
         self.path = settings.get_settings()["megamix_options"]["mod_path"]
-        self.mod_pv = self.path + "/ArchipelagoMod/rom/mod_pv_db.txt"
-        self.songResultsLocation = self.path + "/ArchipelagoMod/results.json"
+        self.mod_name = "ArchipelagoMod"
+        self.mod_pv = f"{self.path}/{self.mod_name}/rom/mod_pv_db.txt"
+        self.songResultsLocation = f"{self.path}/{self.mod_name}/results.json"
         self.modData = None
         self.modded = False
         self.freeplay = False
@@ -370,7 +371,12 @@ class MegaMixContext(CommonContext):
             logger.info("Removed non-AP songs!")
 
     async def restore_songs(self):
-        restore_originals(self.mod_pv_list)
+        mod_pv_dbs = [f"{root}\mod_pv_db.txt" for root, _, files in os.walk(self.path) if 'mod_pv_db.txt' in files]
+        restore_originals(mod_pv_dbs)
+
+    async def shutdown(self):
+        await self.restore_songs()
+        await super().shutdown()
 
 
 def launch():
diff --git a/DataHandler.py b/DataHandler.py
index 67717e2..ea18025 100644
--- a/DataHandler.py
+++ b/DataHandler.py
@@ -1,6 +1,5 @@
 import json
 import yaml
-import pkgutil
 import re
 import os
 import shutil
@@ -8,7 +7,7 @@ import sys
 import settings
 import Utils
 import logging
-from .SymbolFixer import fix_song_name
+import filecmp
 from typing import Any
 
 # Set up logger
@@ -57,8 +56,11 @@ def restore_originals(original_file_paths):
         copy_file_path = os.path.join(directory, copy_filename)
 
         if os.path.exists(copy_file_path):
-            shutil.copyfile(copy_file_path, original_file_path)
-            logger.debug(f"Restored {original_file_path} from {copy_file_path}")
+            if not filecmp.cmp(copy_file_path, original_file_path):
+                shutil.copyfile(copy_file_path, original_file_path)
+                logger.debug(f"Restored {original_file_path} from {copy_file_path}")
+            else:
+                logger.debug(f"Skipping restore on {original_file_path} (matches copy)")
         else:
             logger.debug(f"The copy file {copy_file_path} does not exist.")
 
