diff --git a/Client.py b/Client.py
index 335fd68..726a711 100644
--- a/Client.py
+++ b/Client.py
@@ -27,7 +27,7 @@ from CommonClient import (
     server_loop,
     gui_enabled,
 )
-from NetUtils import NetworkItem, ClientStatus
+from NetUtils import NetworkItem, ClientStatus, Permission
 
 
 class DivaClientCommandProcessor(ClientCommandProcessor):
@@ -278,8 +278,12 @@ class MegaMixContext(CommonContext):
 
     async def end_goal(self):
         message = [{"cmd": "StatusUpdate", "status": ClientStatus.CLIENT_GOAL}]
-        if self.autoRemove:
+
+        if Permission.auto & Permission.from_text(self.permissions.get("release")) == Permission.auto:
+            await self.restore_songs()
+        elif self.autoRemove:
             await self.remove_songs()
+
         await self.send_msgs(message)
 
     async def send_checks(self):
