diff --git a/Client.py b/Client.py
index 9b133ed..e2bfd6c 100644
--- a/Client.py
+++ b/Client.py
@@ -207,11 +207,9 @@ class MegaMixContext(CommonContext):
 
 
     def check_goal(self):
-        if self.leeks_obtained >= self.leeks_needed:
-            if not self.sent_unlock_message:
-                self.sent_unlock_message = True
-                logger.info(f"Got enough leeks! Unlocking goal song: {self.goal_song}")
-
+        if not self.sent_unlock_message and self.leeks_obtained >= self.leeks_needed:
+            self.sent_unlock_message = True
+            logger.info(f"Got enough leeks! Unlocking goal song: {self.goal_song}")
             song_pack = self.is_item_in_modded_data(self.goal_id) if self.modded else "ArchipelagoMod"
             song_unlock(self.path, [self.goal_id], False, song_pack)
 
